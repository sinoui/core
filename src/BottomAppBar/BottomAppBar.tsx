import React from 'react';
import styled, { css } from 'styled-components';
import AppBarActions from '@sinoui/core/AppBarActions';
import NavigationIcon from '@sinoui/core/NavigationIcon';
import InSetCircle from './InSetCircle';

/**
 * 底部应用栏容器
 */
export interface Props {
  /**
   * 浮动操作按钮为居中且嵌入显示
   */
  insertFab?: boolean;
  /**
   * 浮动操作按钮为居右显示 可操作区域要放在左侧显示。
   */
  endFab?: boolean;
  /**
   * 应用栏左侧图标
   */
  navigationIcon?: React.ReactNode;
  /**
   * 浮动操作按钮
   */
  fab?: React.ReactNode;
  /**
   * 应用栏可操作元素
   */
  actionItems?: React.ReactNode;
  style?: React.CSSProperties;
}

const endFabCss = css<Props>`
  && ${AppBarActions} {
    justify-content: ${(props) => props.endFab && 'flex-start'};
  }

  && > .sinoui-bottom-app-bar--fab {
    right: 16px;
  }
`;

/**
 * 非嵌入模式底部应用 容器
 */
const StyledBottomAppBar = styled.div<{
  endFab?: boolean;
}>`
  position: fixed;
  bottom: 0;
  left: 0;
  z-index: ${({ theme }) => theme.zIndex.appBar};
  display: flex;
  height: 56px;
  width: 100%;
  align-items: center;
  padding: 4px;
  box-sizing: border-box;
  color: ${({ theme }) => theme.palette.primary.contrastText};
  box-shadow: ${({ theme }) => theme.shadows[8]};
  background-color: ${(props) => props.theme.palette.primary.main};
  ${(props) => props.endFab && endFabCss}

  & ${AppBarActions} {
    flex: 1;
    justify-content: flex-end;
  }

  & > .sinoui-bottom-app-bar--fab {
    position: absolute;
    right: calc(50% - 28px);
    top: -50%;
  }
`;

/**
 * 嵌入模式底部应用 容器
 */

const InsetBottomAppBar = styled.div`
  position: fixed;
  display: flex;
  bottom: 0;
  left: 0;
  z-index: ${({ theme }) => theme.zIndex.appBar};
  height: 56px;
  padding: 4px;
  width: 100%;
  align-items: center;
  box-sizing: border-box;
  color: ${({ theme }) => theme.palette.primary.contrastText};
  box-shadow: ${({ theme }) => theme.shadows[8]};
  background-color: transparent;

  & ${AppBarActions} {
    flex: 1;
    justify-content: flex-end;
    z-index: 1;
  }

  & ${NavigationIcon} {
    z-index: 1;
  }

  & > .sinoui-bottom-app-bar--fab {
    position: absolute;
    right: calc(50% - 28px);
    top: -50%;
  }
  & .sinoui-fab {
    box-shadow: ${(props) => props.theme.shadows[12]};
  }
`;

const InsetBottomAppBarLayer = styled.div`
  position: absolute;
  display: flex;
  width: 100%;
  height: 56px;
  top: 0;
  left: 0;
  .sinoui-bottom-app-bar--layer {
    flex: 1;
    height: 100%;
    background-color: ${(props) => props.theme.palette.primary.main};
  }
`;

/**
 * 底部应用栏
 */
const BottomAppBar = React.forwardRef<HTMLDivElement, Props>((props, ref) => {
  const { navigationIcon, actionItems, fab, insertFab, style } = props;
  return !insertFab ? (
    <StyledBottomAppBar ref={ref} style={style} endFab={props.endFab}>
      {navigationIcon && <NavigationIcon>{navigationIcon}</NavigationIcon>}
      <div className="sinoui-bottom-app-bar--fab">{fab}</div>
      <AppBarActions>{actionItems}</AppBarActions>
    </StyledBottomAppBar>
  ) : (
    <InsetBottomAppBar style={style}>
      {navigationIcon && <NavigationIcon>{navigationIcon}</NavigationIcon>}
      <div className="sinoui-bottom-app-bar--fab">{fab}</div>
      <AppBarActions>{actionItems}</AppBarActions>
      <InsetBottomAppBarLayer>
        <div className="sinoui-bottom-app-bar--layer" />
        <InSetCircle />
        <div className="sinoui-bottom-app-bar--layer" />
      </InsetBottomAppBarLayer>
    </InsetBottomAppBar>
  );
});

export default BottomAppBar;
