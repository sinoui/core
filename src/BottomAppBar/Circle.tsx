import React, { useContext } from 'react';
import { ThemeContext } from 'styled-components';

export default function InsetCircle() {
  const {
    palette: {
      primary: { main },
    },
  } = useContext(ThemeContext);
  return (
    <svg
      version="1.1"
      x="0px"
      y="0px"
      width="72px"
      height="56px"
      viewBox="0 0 72 56"
      enableBackground="new 0 0 72 56"
    >
      <path
        fill={main}
        d="M36,36.018c-19.882,0-36-16.154-36-36.08v56.125h72V-0.063C72,19.864,55.882,36.018,36,36.018z"
      />
    </svg>
  );
}
