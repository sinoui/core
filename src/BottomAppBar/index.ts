import BottomAppBar from './BottomAppBar';

export default BottomAppBar;
