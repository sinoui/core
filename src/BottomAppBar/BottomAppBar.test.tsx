import React from 'react';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '@sinoui/theme';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import BottomAppBar from '@sinoui/core/BottomAppBar';
import InSetCircle from './InSetCircle';

/**
 * 底部应用栏BottomAppBar组件 单元测试
 */

describe('快照测试', () => {
  it('默认', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <BottomAppBar />
        </ThemeProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('浮动按钮居右显示', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <BottomAppBar endFab />
        </ThemeProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('浮动按钮居中嵌入', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <BottomAppBar insertFab />
        </ThemeProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('嵌入图标', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <InSetCircle />
        </ThemeProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
