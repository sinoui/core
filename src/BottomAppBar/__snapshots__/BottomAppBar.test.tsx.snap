// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`快照测试 嵌入图标 1`] = `
<svg
  enableBackground="new 0 0 72 56"
  height="56px"
  version="1.1"
  viewBox="0 0 72 56"
  width="72px"
  x="0px"
  y="0px"
>
  <path
    d="M36,36.018c-19.882,0-36-16.154-36-36.08v56.125h72V-0.063C72,19.864,55.882,36.018,36,36.018z"
    fill="#3f51b5"
  />
</svg>
`;

exports[`快照测试 浮动按钮居中嵌入 1`] = `
<div
  className="BottomAppBar__InsetBottomAppBar-f0o39x-1 jemRnS"
>
  <div
    className="sinoui-bottom-app-bar--fab"
  />
  <div
    className="AppBarActions-sl4kx3-0 euUhBN sinoui-app-bar--actions"
  />
  <div
    className="BottomAppBar__InsetBottomAppBarLayer-f0o39x-2 leNtDb"
  >
    <div
      className="sinoui-bottom-app-bar--layer"
    />
    <svg
      enableBackground="new 0 0 72 56"
      height="56px"
      version="1.1"
      viewBox="0 0 72 56"
      width="72px"
      x="0px"
      y="0px"
    >
      <path
        d="M36,36.018c-19.882,0-36-16.154-36-36.08v56.125h72V-0.063C72,19.864,55.882,36.018,36,36.018z"
        fill="#3f51b5"
      />
    </svg>
    <div
      className="sinoui-bottom-app-bar--layer"
    />
  </div>
</div>
`;

exports[`快照测试 浮动按钮居右显示 1`] = `
<div
  className="BottomAppBar__StyledBottomAppBar-f0o39x-0 koyuBh"
>
  <div
    className="sinoui-bottom-app-bar--fab"
  />
  <div
    className="AppBarActions-sl4kx3-0 euUhBN sinoui-app-bar--actions"
  />
</div>
`;

exports[`快照测试 默认 1`] = `
<div
  className="BottomAppBar__StyledBottomAppBar-f0o39x-0 gengsm"
>
  <div
    className="sinoui-bottom-app-bar--fab"
  />
  <div
    className="AppBarActions-sl4kx3-0 euUhBN sinoui-app-bar--actions"
  />
</div>
`;
